src/music/midifile.cpp

```diff
@@ -450,8 +794,304 @@ void MidiFile::MoveFrom(MidiFile &other)
+		fwrite(tb, 1, sizeof(tb), f);
+	}
+}
+
+/**
+ * Write a Standard MIDI File containing the decoded music.
+ * @param filename Name of file to write to
+ * @return True if the file was written to completion
+ */
+bool MidiFile::WriteSMF(const char *filename)
+{
+	FILE *f = FioFOpenFile(filename, "wb", Subdirectory::NO_DIRECTORY);
+	if (!f) {
+		return false;
+	}
+
+	uint16 u16;
+	byte bb;
+
+	/* SMF header */
+	const byte filemagic[] = { 'M', 'T', 'h', 'd', 0x00, 0x00, 0x00, 0x06 };
+	fwrite(filemagic, sizeof(filemagic), 1, f);
+	fwrite(&(u16 = 0), sizeof(u16), 1, f);
+	fwrite(&(u16 = TO_BE16(1)), sizeof(u16), 1, f);
+	fwrite(&(u16 = TO_BE16(this->tickdiv)), sizeof(u16), 1, f);
+
+	/* Track header, block length is written after everything
+	 * else has already been written and length is known. */
+	const byte trackmagic[] = { 'M', 'T', 'r', 'k', 0, 0, 0, 0 };
+	fwrite(trackmagic, sizeof(trackmagic), 1, f);
+	size_t tracksizepos = ftell(f) - 4;
+
+	/* Write blocks in sequence */
+	uint32 lasttime = 0;
+	size_t nexttempoindex = 0;
+	for (size_t bi = 0; bi < this->blocks.size(); bi++) {
+	redoblock:
+		DataBlock &block = this->blocks[bi];
+		TempoChange &nexttempo = this->tempos[nexttempoindex];
+
+		uint32 timediff = block.ticktime - lasttime;
+
+		/* Check if there is a tempo change before this block */
+		if (nexttempo.ticktime < block.ticktime) {
+			timediff = nexttempo.ticktime - lasttime;
+		}
+
+		/* Write delta time for block */
+		lasttime += timediff;
+		bool needtime = false;
+		WriteVariableLen(f, timediff);
+
+		/* Write tempo change if there is one */
+		if (nexttempo.ticktime <= block.ticktime) {
+			byte tempobuf[6] = { MIDIST_SMF_META, 0x51, 0x03, 0, 0, 0 };
+			tempobuf[3] = (nexttempo.tempo & 0x00FF0000) >> 16;
+			tempobuf[4] = (nexttempo.tempo & 0x0000FF00) >>  8;
+			tempobuf[5] = (nexttempo.tempo & 0x000000FF);
+			fwrite(tempobuf, 1, 6, f);
+			nexttempoindex++;
+			needtime = true;
+		}
+		/* If a tempo change occurred between two blocks, rather than
+		 * at start of this one, start over with delta time for the block. */
+		if (nexttempo.ticktime < block.ticktime) {
+			goto redoblock;
+		}
+
+		/* Write each block data command */
+		byte *dp = block.data.Begin();
+		while (dp < block.data.End()) {
+			/* Always zero delta time inside blocks */
+			if (needtime) {
+				fwrite(&(bb = 0), 1, 1, f);
+			}
+			needtime = true;
+
+			/* Check message type and write appropriate number of bytes */
+			switch (*dp & 0xF0) {
+			case MIDIST_NOTEOFF:
+			case MIDIST_NOTEON:
+			case MIDIST_POLYPRESS:
+			case MIDIST_CONTROLLER:
+			case MIDIST_PITCHBEND:
+				fwrite(dp, 1, 3, f);
+				dp += 3;
+				continue;
+			case MIDIST_PROGCHG:
+			case MIDIST_CHANPRESS:
+				fwrite(dp, 1, 2, f);
+				dp += 2;
+				continue;
+			}
+
+			/* Sysex needs to measure length and write that as well */
+			if (*dp == MIDIST_SYSEX) {
+				fwrite(dp, 1, 1, f);
+				dp++;
+				byte *sysexend = dp;
+				while (*sysexend++ != MIDIST_ENDSYSEX);
```

Comment:
`while ();` always looks like a mistake - either add a `{}` or put the actual change into the body